api/src/main/java/com/spotify/ffwd/util/HighFrequencyDetector.java

```diff
@@ -0,0 +1,167 @@
+/**
+ * Class responsible for high frequency metrics detection.
+ * All batched metrics will be examined.
+ * If configured will drop metrics marked as high frequency.
+ */
+public class HighFrequencyDetector {
+
+  /** Allow to drop high frequency metrics. */
+  @Inject
+  @Named("dropHighFrequencyMetric")
+  boolean dropHighFrequencyMetric;
+
+  /** Minimum number of milliseconds allowed between data points. */
+  @Inject
+  @Named("minFrequencyMillisAllowed")
+  int minFrequencyMillisAllowed;
+
+  /** Minimum number of times high frequency detected before metrics are dropped. */
+  @Inject
+  @Named("minNumberOfTriggers")
+  int minNumberOfTriggers;
+
+  /** Milliseconds high frequency triggers are refreshed. */
+  @Inject
+  @Named("highFrequencyDataRecycleMS")
+  long highFrequencyDataRecycleMS;
+
+  @Inject
+  private HighFrequencyDetectorStatistics statistics;
+
+  /** High frequency metrics counter. */
+  // TODO review to use ConcurrentHashMap vs AtomicReference
+  final AtomicReference<Map<Integer, Integer>> highFrequencyMetrics =
+      new AtomicReference<>(new HashMap<>());
+
+  final AtomicLong highFrequencyTriggersTS;
+  private final Object lock = new Object();
+
+  /* TODO
+      1. unit tests
+          a. sendBatch with many points
+          b. send both with many points
+          c. check swap of high freq hashmap
+   */
+
+  @Inject
+  public HighFrequencyDetector() {
+    this.highFrequencyTriggersTS = new AtomicLong(System.currentTimeMillis());
+  }
+
+  /**
+   * Detects high frequency metrics by grouping and calculating
+   * time delta between metric timestamps
+   *
+   * @param metrics
+   * @return list of filtered metrics
+   */
+  public List<Metric> detect(final List<Metric> metrics) {
+    List<Metric> newList = new ArrayList<>();
+
+    // Groups metrics by metric hash code and finds times deltas of ordered data points.
+    // {hashcode -> -1, hashcode1 -> 10}
+    Map<Integer, Integer> groupedMetrics =
+        metrics.stream()
+            .sorted(Comparator.comparing(Metric::getTime))
+            .collect(
+                Collectors.groupingBy(
+                    Metric::hashCode,
+                    Collectors.collectingAndThen(
+                        Collectors.toList(),
+                        this::computeTimeDelta)));
+
+    updateHighFrequencyMetricsStats(groupedMetrics);
+
+    if (dropHighFrequencyMetric && highFrequencyMetrics.get().size() > 0) {
+      metrics.stream()
+          .filter(
+              metric ->
+                  highFrequencyMetrics.get().getOrDefault(metric.hashCode(), 0)
+                      < minNumberOfTriggers)
+          .forEach(newList::add);
+
+      statistics.reportHighFrequencyMetricsDropped(metrics.size() - newList.size());
+      return newList;
+    }
+
+    return metrics;
+  }
+
+  private int computeTimeDelta(List<Metric> list) {
+    int size = list.size();
+    return IntStream.range(1, size)
+        .map(
+            x ->
+                (int)
```

Comment:
May be I am missing something :). I don't see the point of casting.  Even if the difference is small, it is good practice to avoid this type of casting.

Reply:
Casting needed as the computations done on `long` values and if I remove casting I'm getting this

`Bad return type in lambda expression: long cannot be converted to int`

Reply:
I think this is because of the method return type which is  an int.